org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/debug/FqnDiscover.java

```diff
@@ -0,0 +1,311 @@
+
+    if (javaProject == null) {
+      throw new IllegalArgumentException(format("Project for '%s' not found", fileUri));
+    }
+
+    URI fileRelativeUri =
+        javaProject.getProject().getLocationURI().relativize(JDTUtils.toURI(fileUri));
+    IPath fileRelativePath = javaProject.getPath().append(fileRelativeUri.toString());
+
+    String fqn = null;
+    for (int i = fileRelativePath.segmentCount(); i > 0; i--) {
+      try {
+        IClasspathEntry classpathEntry =
+            ((JavaProject) javaProject)
+                .getClasspathEntryFor(fileRelativePath.removeLastSegments(i));
+
+        if (classpathEntry != null) {
+          fqn =
+              fileRelativePath
+                  .removeFirstSegments(fileRelativePath.segmentCount() - i)
+                  .removeFileExtension()
+                  .toString()
+                  .replace("/", ".");
+          break;
+        }
+      } catch (JavaModelException e) {
+        return fileUri;
+      }
+    }
+
+    if (fqn == null) {
+      return fileUri;
+    }
+
+    IType outerClass;
+    IJavaElement iMember;
+    try {
+      outerClass = javaProject.findType(fqn);
+
+      if (outerClass == null) {
+        return fileUri;
+      }
+
+      String source;
+      if (outerClass.isBinary()) {
+        IClassFile classFile = outerClass.getClassFile();
+        source = classFile.getSource();
+      } else {
+        ICompilationUnit unit = outerClass.getCompilationUnit();
+        source = unit.getSource();
+      }
+
+      Document document = new Document(source);
+      IRegion region = document.getLineInformation(lineNumber);
+      int start = region.getOffset();
+      int end = start + region.getLength();
+
+      iMember = binSearch(outerClass, start, end);
+    } catch (JavaModelException | BadLocationException e) {
+      throw new IllegalArgumentException(e);
+    }
+
+    if (iMember instanceof IType) {
+      return ((IType) iMember).getFullyQualifiedName();
+    }
+
+    if (iMember != null) {
+      fqn = ((IMember) iMember).getDeclaringType().getFullyQualifiedName();
+
+      // figure out if the fqn represents a local inner class
+      while (iMember != null && !(iMember instanceof CompilationUnit)) {
+        if (iMember instanceof SourceType
+            && !((SourceType) iMember).isAnonymous()
+            && iMember.getParent() instanceof SourceMethod) {
+
+          fqn = fqn.replace("$" + iMember.getElementName(), "$1" + iMember.getElementName());
+        }
+
+        iMember = iMember.getParent();
+      }
+
+      return fqn;
+    }
+
+    throw new IllegalArgumentException(
+        format("Unable to find source code for '%s' in the project '%s'", fileUri, javaProject));
+  }
+
+  /**
+   * Finds resources by a fqn.
+   *
+   * @param params contains fqn
+   * @return all resources are identified by the given fqn
+   */
+  public static List<ResourceLocationParameters> findResourcesByFqn(
```

Comment:
If it's used as a result, it shouldn't be names "...Parameters". Why not just ResourceLocation?